# Breast Cancer Prediction 

---

---

## Contents

1. [Background](#Background)
1. [Setup](#Setup)
1. [Data](#Data)
1. [Train](#Train)

---

## Background
For this illustration, we have taken an example for breast cancer prediction using UCI'S breast cancer diagnostic data set available at https://archive.ics.uci.edu/ml/datasets/Breast+Cancer+Wisconsin+%28Diagnostic%29. The data set is also available on Kaggle at https://www.kaggle.com/uciml/breast-cancer-wisconsin-data. The purpose here is to use this data set to build a predictve model of whether a breast mass image indicates benign or malignant tumor. 

---

## Setup

Now we'll import the Python libraries we'll need.

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import io
import time
import json

import al

---
## Data

Data Source: https://archive.ics.uci.edu/ml/machine-learning-databases/breast-cancer-wisconsin/wdbc.data
        https://www.kaggle.com/uciml/breast-cancer-wisconsin-data

Let's download the data and save it in the local folder with the name data.csv and take a look at it.

In [3]:
data = pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/breast-cancer-wisconsin/wdbc.data', header = None)

# specify columns extracted from wbdc.names
data.columns = ["id","diagnosis","radius_mean","texture_mean","perimeter_mean","area_mean","smoothness_mean",
                "compactness_mean","concavity_mean","concave points_mean","symmetry_mean","fractal_dimension_mean",
                "radius_se","texture_se","perimeter_se","area_se","smoothness_se","compactness_se","concavity_se",
                "concave points_se","symmetry_se","fractal_dimension_se","radius_worst","texture_worst",
                "perimeter_worst","area_worst","smoothness_worst","compactness_worst","concavity_worst",
                "concave points_worst","symmetry_worst","fractal_dimension_worst"] 

# save the data
data.to_csv("data.csv", sep=',', index=False)

# print the shape of the data file
print(data.shape)

# show the top few rows
display(data.head())

# describe the data object
display(data.describe())

# we will also summarize the categorical field diganosis 
display(data.diagnosis.value_counts())


(569, 32)


id diagnosis  radius_mean  texture_mean  perimeter_mean  area_mean  \
0    842302         M        17.99         10.38          122.80     1001.0   
1    842517         M        20.57         17.77          132.90     1326.0   
2  84300903         M        19.69         21.25          130.00     1203.0   
3  84348301         M        11.42         20.38           77.58      386.1   
4  84358402         M        20.29         14.34          135.10     1297.0   

   smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
0          0.11840           0.27760          0.3001              0.14710   
1          0.08474           0.07864          0.0869              0.07017   
2          0.10960           0.15990          0.1974              0.12790   
3          0.14250           0.28390          0.2414              0.10520   
4          0.10030           0.13280          0.1980              0.10430   

   ...  radius_worst  texture_worst  perimeter_worst  area_worst  \
0  ...         25.38          17.33           184.60      2019.0   
1  ...         24.99          23.41           158.80      1956.0   
2  ...         23.57          25.53           152.50      1709.0   
3  ...         14.91          26.50            98.87       567.7   
4  ...         22.54          16.67           152.20      1575.0   

   smoothness_worst  compactness_worst  concavity_worst  concave points_worst  \
0            0.1622             0.6656           0.7119                0.2654   
1            0.1238             0.1866           0.2416                0.1860   
2            0.1444             0.4245           0.4504                0.2430   
3            0.2098             0.8663           0.6869                0.2575   
4            0.1374             0.2050           0.4000                0.1625   

   symmetry_worst  fractal_dimension_worst  
0          0.4601                  0.11890  
1          0.2750                  0.08902  
2          0.3613                  0.08758  
3          0.6638                  0.17300  
4          0.2364                  0.07678  

[5 rows x 32 columns]

id  radius_mean  texture_mean  perimeter_mean    area_mean  \
count  5.690000e+02   569.000000    569.000000      569.000000   569.000000   
mean   3.037183e+07    14.127292     19.289649       91.969033   654.889104   
std    1.250206e+08     3.524049      4.301036       24.298981   351.914129   
min    8.670000e+03     6.981000      9.710000       43.790000   143.500000   
25%    8.692180e+05    11.700000     16.170000       75.170000   420.300000   
50%    9.060240e+05    13.370000     18.840000       86.240000   551.100000   
75%    8.813129e+06    15.780000     21.800000      104.100000   782.700000   
max    9.113205e+08    28.110000     39.280000      188.500000  2501.000000   

       smoothness_mean  compactness_mean  concavity_mean  concave points_mean  \
count       569.000000        569.000000      569.000000           569.000000   
mean          0.096360          0.104341        0.088799             0.048919   
std           0.014064          0.052813        0.079720             0.038803   
min           0.052630          0.019380        0.000000             0.000000   
25%           0.086370          0.064920        0.029560             0.020310   
50%           0.095870          0.092630        0.061540             0.033500   
75%           0.105300          0.130400        0.130700             0.074000   
max           0.163400          0.345400        0.426800             0.201200   

       symmetry_mean  ...  radius_worst  texture_worst  perimeter_worst  \
count     569.000000  ...    569.000000     569.000000       569.000000   
mean        0.181162  ...     16.269190      25.677223       107.261213   
std         0.027414  ...      4.833242       6.146258        33.602542   
min         0.106000  ...      7.930000      12.020000        50.410000   
25%         0.161900  ...     13.010000      21.080000        84.110000   
50%         0.179200  ...     14.970000      25.410000        97.660000   
75%         0.195700  ...     18.790000      29.720000       125.400000   
max         0.304000  ...     36.040000      49.540000       251.200000   

        area_worst  smoothness_worst  compactness_worst  concavity_worst  \
count   569.000000        569.000000         569.000000       569.000000   
mean    880.583128          0.132369           0.254265         0.272188   
std     569.356993          0.022832           0.157336         0.208624   
min     185.200000          0.071170           0.027290         0.000000   
25%     515.300000          0.116600           0.147200         0.114500   
50%     686.500000          0.131300           0.211900         0.226700   
75%    1084.000000          0.146000           0.339100         0.382900   
max    4254.000000          0.222600           1.058000         1.252000   

       concave points_worst  symmetry_worst  fractal_dimension_worst  
count            569.000000      569.000000               569.000000  
mean               0.114606        0.290076                 0.083946  
std                0.065732        0.061867                 0.018061  
min                0.000000        0.156500                 0.055040  
25%                0.064930        0.250400                 0.071460  
50%                0.099930        0.282200                 0.080040  
75%                0.161400        0.317900                 0.092080  
max                0.291000        0.663800                 0.207500  

[8 rows x 31 columns]

B    357
M    212
Name: diagnosis, dtype: int64

#### Key observations:
* Data has 569 observations and 32 columns.
* First field is 'id'.
* Second field, 'diagnosis', is an indicator of the actual diagnosis ('M' = Malignant; 'B' = Benign).
* There are 30 other numeric features available for prediction.

## Create Features and Labels
#### Split the data into 80% training, 10% validation and 10% testing.

In [5]:
y = ((data.iloc[:,1] == 'M') +0).values;
X = data.iloc[:,2:].values;

---
## Train


In [8]:
m = al.AL()

m.fit(X, y)
progs = m.get_programs()

  0%|          | 0/26 [00:00<?, ?it/s]

Generating transforms
Programs of depth=0


Failed to apply transform(sklearn.pipeline.Pipeline): None
Failed to apply transform(sklearn.pipeline.Pipeline): <class 'synthesis.runtime_helpers.ColumnLoop'>
Failed to apply transform(builtins.module): None
Failed to apply transform(builtins.module): <class 'synthesis.runtime_helpers.ColumnLoop'>
Failed to apply transform(sklearn.feature_extraction.dict_vectorizer.DictVectorizer): None
Failed to apply transform(sklearn.feature_extraction.dict_vectorizer.DictVectorizer): <class 'synthesis.runtime_helpers.ColumnLoop'>
Failed to apply transform(sklearn.feature_extraction.text.CountVectorizer): None
Failed to apply transform(sklearn.feature_extraction.text.CountVectorizer): <class 'synthesis.runtime_helpers.ColumnLoop'>

 19%|█▉        | 5/26 [00:00<00:00, 40.53it/s]Failed to apply transform(sklearn.ensemble.forest.RandomForestRegressor): None
Failed to apply transform(sklearn.ensemble.forest.RandomForestRegressor): <class 'synthesis.runtime_helpers.ColumnLoop'>
Failed to apply transform

[Transforms] Pruning programs of depth 0
Programs of depth=1


Failed to apply transform(sklearn.pipeline.Pipeline): None
Failed to apply transform(sklearn.pipeline.Pipeline): <class 'synthesis.runtime_helpers.ColumnLoop'>
Failed to apply transform(builtins.module): None
Failed to apply transform(builtins.module): <class 'synthesis.runtime_helpers.ColumnLoop'>
Failed to apply transform(sklearn.feature_extraction.dict_vectorizer.DictVectorizer): None
Failed to apply transform(sklearn.feature_extraction.dict_vectorizer.DictVectorizer): <class 'synthesis.runtime_helpers.ColumnLoop'>

 19%|█▉        | 5/26 [00:00<00:00, 45.92it/s]Failed to apply transform(sklearn.feature_extraction.text.CountVectorizer): None
Failed to apply transform(sklearn.feature_extraction.text.CountVectorizer): <class 'synthesis.runtime_helpers.ColumnLoop'>
Failed to apply transform(sklearn.decomposition.truncated_svd.TruncatedSVD): <class 'synthesis.runtime_helpers.ColumnLoop'>

 31%|███       | 8/26 [00:00<00:00, 34.41it/s]Failed to apply transform(sklearn.ensemble.forest.Rand

Failed computing missing


Failed to apply transform(sklearn.feature_extraction.dict_vectorizer.DictVectorizer): None
Failed to apply transform(sklearn.feature_extraction.dict_vectorizer.DictVectorizer): <class 'synthesis.runtime_helpers.ColumnLoop'>
Failed to apply transform(builtins.module): None
Failed to apply transform(builtins.module): <class 'synthesis.runtime_helpers.ColumnLoop'>
Failed to apply transform(sklearn.feature_extraction.text.CountVectorizer): None
Failed to apply transform(sklearn.feature_extraction.text.CountVectorizer): <class 'synthesis.runtime_helpers.ColumnLoop'>
Failed to apply transform(sklearn.feature_extraction.text.TfidfVectorizer): None
Failed to apply transform(sklearn.feature_extraction.text.TfidfVectorizer): <class 'synthesis.runtime_helpers.ColumnLoop'>

 15%|█▌        | 4/26 [00:00<00:00, 28.53it/s]Failed to apply transform(sklearn.decomposition.truncated_svd.TruncatedSVD): <class 'synthesis.runtime_helpers.ColumnLoop'>
Failed to apply transform(sklearn.preprocessing.imputatio

[Transforms] Pruning programs of depth 1
Programs of depth=2


Failed to apply transform(sklearn.pipeline.Pipeline): <class 'synthesis.runtime_helpers.ColumnLoop'>
Failed to apply transform(builtins.module): None
Failed to apply transform(builtins.module): <class 'synthesis.runtime_helpers.ColumnLoop'>
Failed to apply transform(sklearn.feature_extraction.text.CountVectorizer): None
Failed to apply transform(sklearn.feature_extraction.text.CountVectorizer): <class 'synthesis.runtime_helpers.ColumnLoop'>
Failed to apply transform(sklearn.feature_extraction.dict_vectorizer.DictVectorizer): None
Failed to apply transform(sklearn.feature_extraction.dict_vectorizer.DictVectorizer): <class 'synthesis.runtime_helpers.ColumnLoop'>
Failed to apply transform(sklearn.feature_selection.from_model.SelectFromModel): None
Failed to apply transform(sklearn.feature_selection.from_model.SelectFromModel): <class 'synthesis.runtime_helpers.ColumnLoop'>
Failed to apply transform(sklearn.decomposition.truncated_svd.TruncatedSVD): <class 'synthesis.runtime_helpers.Column

[Transforms] Pruning programs of depth 2
Generating modeling


Failed to apply model: builtins.module
Failed to apply model: xgboost.core.Booster

  0%|          | 0/32 [00:00<?, ?it/s]/home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages/sklearn/neural_network/multilayer_perceptron.py:563: ConvergenceWarning: Stochastic Optimizer: Maximum iterations reached and the optimization hasn't converged yet.
  % (), ConvergenceWarning)

  3%|▎         | 1/32 [00:00<00:14,  2.07it/s]Failed to apply model: builtins.module
Failed to apply model: xgboost.core.Booster

  0%|          | 0/32 [00:00<?, ?it/s]/home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages/sklearn/neural_network/multilayer_perceptron.py:563: ConvergenceWarning: Stochastic Optimizer: Maximum iterations reached and the optimization hasn't converged yet.
  % (), ConvergenceWarning)

  3%|▎         | 1/32 [00:00<00:15,  2.07it/s]Failed to apply model: builtins.module

 22%|██▏       | 7/32 [00:00<00:02, 11.91it/s]Failed to apply model: xgboost.core.Booster

 16%|█▌      

Failed computing missing



 19%|█▉        | 6/32 [00:00<00:02,  9.11it/s]Failed to apply model: builtins.module
Failed to apply model: sklearn.neighbors.kde.KernelDensity

 31%|███▏      | 10/32 [00:01<00:02,  8.93it/s]Failed to apply model: xgboost.core.Booster

  0%|          | 0/32 [00:00<?, ?it/s]Failed to apply model: builtins.module
Failed to apply model: xgboost.core.Booster

 31%|███▏      | 10/32 [00:00<00:01, 14.46it/s]Failed to apply model: sklearn.grid_search.GridSearchCV

  0%|          | 0/32 [00:00<?, ?it/s]Failed to apply model: builtins.module
Failed to apply model: xgboost.core.Booster

 31%|███▏      | 10/32 [00:00<00:01, 14.63it/s]Failed to apply model: sklearn.grid_search.GridSearchCV

  0%|          | 0/32 [00:00<?, ?it/s]Failed to apply model: builtins.module
Failed to apply model: xgboost.core.Booster

 16%|█▌        | 5/32 [00:00<00:02, 10.78it/s]Failed to apply model: builtins.module

 22%|██▏       | 7/32 [00:00<00:02, 11.73it/s]Failed to apply model: xgboost.core.Booster

 16%|█▌    

[Model] Pruning programs of depth 0
[Model] Pruning programs of depth 1
[Model] Pruning programs of depth 2
[Model] Pruning programs of depth 3
Adding test data


Failed to add test data, program depth: 0
/home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1113: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
 15%|█▌        | 6/40 [00:00<00:00, 54.31it/s]/home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages/sklearn/metrics/classification.py:1113: UndefinedMetricWarning: F-score is ill-defined and being set to 0.0 in labels with no predicted samples.
  'precision', 'predicted', average, warn_for)
 30%|███       | 12/40 [00:00<00:00, 53.34it/s]Failed to add test data, program depth: 1
Failed to add test data, program depth: 1
 75%|███████▌  | 30/40 [00:00<00:00, 40.51it/s]Failed to add test data, program depth: 3
Failed to add test data, program depth: 3
  0%|          | 0/34 [00:00<?, ?it/s]/home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages/sklearn/neural_network/multila

Print pipline code

In [9]:
print(progs[0].pipeline_code())
print('\n\n-----------------------\n')
print(progs[1].pipeline_code())
print('\n\n-----------------------\n')
print(progs[3].pipeline_code())
print('\n\n-----------------------\n')
print(progs[4].pipeline_code())

import sklearn.neural_network.multilayer_perceptron
import sklearn.preprocessing.imputation
import xgboost
import sklearn.preprocessing.data
import sklearn

from sklearn.pipeline import Pipeline
p = Pipeline([('t0', runtime_helpers.ColumnLoop(sklearn.preprocessing.data.StandardScaler)),('t1', sklearn.preprocessing.imputation.Imputer()),('t2', sklearn.preprocessing.data.MinMaxScaler()),('model', sklearn.neural_network.multilayer_perceptron.MLPClassifier())])
p.fit(X_train, y_train)
print(p.score(X_val, y_val))
p.fit(X, y)
def predict(X_test): return p.predict(X_test)


-----------------------

import sklearn.neural_network.multilayer_perceptron
import sklearn.preprocessing.imputation
import xgboost
import sklearn.preprocessing.data
import sklearn

from sklearn.pipeline import Pipeline
p = Pipeline([('t0', runtime_helpers.ColumnLoop(sklearn.preprocessing.data.StandardScaler)),('t1', sklearn.preprocessing.imputation.Imputer()),('t2', runtime_helpers.ColumnLoop(sklearn.preprocessing.data.M

In [20]:
import synthesis.runtime_helpers as runtime_helpers
from sklearn.model_selection import train_test_split

X_train, X_val, y_train, y_val = train_test_split(X, y, test_size=0.2)

Select pipeline and paste it's code below

In [21]:
import sklearn.neural_network.multilayer_perceptron
import xgboost
import sklearn.preprocessing.data
import sklearn

from sklearn.pipeline import Pipeline
p = Pipeline([('t0', runtime_helpers.ColumnLoop(sklearn.preprocessing.data.StandardScaler)),('model', sklearn.neural_network.multilayer_perceptron.MLPClassifier())])
p.fit(X_train, y_train)
print(p.score(X_val, y_val))
p.fit(X, y)
def predict(X_test): return p.predict(X_test)

/home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages/sklearn/neural_network/multilayer_perceptron.py:563: ConvergenceWarning: Stochastic Optimizer: Maximum iterations reached and the optimization hasn't converged yet.
  % (), ConvergenceWarning)


0.991228070175


/home/ec2-user/anaconda3/envs/python3/lib/python3.6/site-packages/sklearn/neural_network/multilayer_perceptron.py:563: ConvergenceWarning: Stochastic Optimizer: Maximum iterations reached and the optimization hasn't converged yet.
  % (), ConvergenceWarning)
